### ISSUE TITLE
Phylogenetic inference via Bethe free energy models

### ISSUE THREAD

**Issue by fritzo** [broadinstitute/pyro-cov#4]
This issue tracks progress towards end-to-end phylogenetic inference in Pyro using Bethe free energy approximations and a relaxed genetic embedding.

## Model improvements
- [ ] switch from mean-field to joint relaxation via either Gumbel-max or [torch.cdist](https://pytorch.org/docs/stable/generated/torch.cdist.html#torch.cdist)
- [ ] try a multi-layer decoder
- [ ] try the hyperbolic decoder of [Mathieu et al. (2019)](https://papers.nips.cc/paper/2019/file/0ec04cb3912c4f08874dd03716f80df1-Paper.pdf)
- [ ] try hyperbolic normalizing flows of [Bose et al. (2020)](https://arxiv.org/abs/2002.06336)
- [ ] try extra internal nodes as in a multi-bernoulli random field. see the [binary-matching branch](https://github.com/pyro-ppl/pyro/tree/binary-matching)
- [ ] try a determinantal point process as in [Kulesza & Taskar (2012)](https://arxiv.org/abs/1207.6083)

## Inference components
- [x] `OneTwoMatching.log_partition_function` via Sinkhorn + Bethe approximation
- [x] `nn.Linear` decoder
- [x] SVD initialization of latent codes
- [ ] bijective `CoalescentTransform` for the `CoalescentConstraint` with heterochronous leaves
- [ ] use implicit differentiation in the Sinkhorn loop, as DEQ models ([paper](https://papers.nips.cc/paper/2019/file/01386bd6d8e091c2ab4c7c7de644d37b-Paper.pdf) | [code](https://github.com/locuslab/deq))
- [ ] replace Sinkhorn with an optimal solver
- [ ] `OneTwoMatching.sample()`, say via purturb-and-map random fields or quickly converging MCMC

## Validation plan
- [x] unit test `.log_partition_function` on tiny data
- [x] qualitatively match skyline plots from Beast
- [ ] quantitatively match skyline plots from Beast (requires matching coalescent & substitution models)
- [ ] Compare trees inferred with Beast on datasets DS1-DS11 of [Whidden & Matsen (2015)](https://academic.oup.com/sysbio/article/64/3/472/1632660)
- [ ] compare SVI vs MCMC for the Bethe model
- [ ] quantify compute cost on large data (10k, 100k taxa)
- [ ] unit test `.log_partition_function` on top say 1000 samples from BEAST on medium data (<100 taxa)
- [ ] unit test `.log_partition_function` via thermodynamic annealing on medium-large data (1k taxa)

## Theory & writeup
- [x] Determine whether Sinkhorn is optimal; possibly compare with optimal solution.
- [x] Determine whether the relaxed genetic embedding still permits a lower bound.
